Wikipedia/Code/WMFArticleImageProtocol.m

```diff
@@ -55,12 +71,44 @@ - (void)startLoading {
     DDLogVerbose(@"Fetching image %@", self.request.URL);
     @weakify(self);
     [[WMFImageController sharedInstance] fetchImageWithURL:self.request.URL]
-    .then(^(WMFImageDownload* download) {
+    .thenInBackground(^WMFArticleImageProtocolResponseAndData*(WMFImageDownload* download) {
         @strongify(self);
         if(!self){
-            return;
+            return nil;
         }
-        [self respondWithDataFromDownload:download];
+        return [self getResponseAndDataForDownload:download];
+    })
+    .then(^id(WMFArticleImageProtocolResponseAndData* responseAndData) {
+        @strongify(self);
+        if(!self){
+            return nil;
+        }
+
+        if (!responseAndData.response) {
+            return [[NSError alloc] initWithDomain:WMFArticleImageProtocolResponseAndDataErrorDomain
+                                              code:0
+                                          userInfo:@{
+                                                     NSLocalizedDescriptionKey: @"Response for image not found",
+                                                     NSURLErrorKey: self.request.URL
+                                                     }];
+        }else if (!responseAndData.data) {
```

Comment:
You can remove this check and just use the one below… `responseAndData.data.length=0` will catch this as well.